File: gem/socket-patch/lib/socket_patch/launcher.rb
# frozen_string_literal: true

require "rbconfig"
require "digest"
require "fileutils"
require "net/http"
require "uri"
require "tmpdir"

module SocketPatch
  # Resolves and runs the prebuilt `socket-patch` binary for the host platform.
  #
  # Strategy (mirrors scripts/install.sh's target mapping):
  #   1. honor SOCKET_PATCH_BIN if it points at an executable (airgap escape);
  #   2. else use a cached binary under the per-user cache, keyed by
  #      version + target;
  #   3. else download `socket-patch-<target>.{tar.gz,zip}` from the matching
  #      GitHub release, verify its SHA-256 against the release's SHA256SUMS,
  #      extract the binary, cache it, and run it.
  module Launcher
    # Fallback version, used ONLY when the installed gem's version can't be read
    # (e.g. running this file from a checkout). In a real `gem install` the
    # download uses the installed gem's own version — see `version`.
    VERSION = "3.3.0"
    REPO = "SocketDev/socket-patch"
    BINARY = "socket-patch"

    module_function

    def run(argv)
      bin = resolve_binary
      if Gem.win_platform?
        # Windows has no exec() that replaces the process cleanly for console
        # apps; spawn + wait and propagate the child's exit status.
        exit(system(bin, *argv) ? $?.exitstatus : 1)
      else
        exec([bin, bin], *argv)
      end
    rescue LauncherError => e
      warn("socket-patch: #{e.message}")
      exit(1)
    end

    class LauncherError < StandardError; end

    # ── binary resolution ─────────────────────────────────────────────────────

    def resolve_binary
      env = ENV["SOCKET_PATCH_BIN"]
      return env if env && !env.empty? && File.executable?(env)

      ver = version
      target, ext = detect_target
      exe = BINARY + (Gem.win_platform? ? ".exe" : "")
      cached = File.join(cache_dir, ver, target, exe)
      return cached if File.executable?(cached)

      download_binary(ver, target, ext, cached)
      cached
    end

    # The version to fetch — the binary MUST match the CLI package the user
    # actually installed, so derive it from the installed gem's own spec rather
    # than trusting the `VERSION` constant (which `version-sync.sh` keeps current
    # but which could drift). Falls back to the constant when the gem isn't
    # activated (e.g. running this file directly from a checkout).
    def version
      if (spec = Gem.loaded_specs["socket-patch"])
        return spec.version.to_s
      end
      Gem::Specification.find_by_name("socket-patch").version.to_s
    rescue StandardError
      VERSION
    end

    # Map the host to a release target triple + archive extension. Mirrors
    # scripts/install.sh.
    def detect_target
      host_os = RbConfig::CONFIG["host_os"].downcase
      host_cpu = RbConfig::CONFIG["host_cpu"].downcase

      arch =
        case host_cpu
        when /x86_64|x64|amd64/ then "x86_64"
        when /aarch64|arm64/ then "aarch64"
        when /i[3-6]86|x86/ then "i686"
        when /armv7|armhf|arm\b/ then "arm"
        else raise LauncherError, "unsupported CPU architecture: #{host_cpu}"
        end

      case host_os
      when /darwin|mac/
        raise LauncherError, "unsupported macOS arch: #{arch}" unless %w[x86_64 aarch64].include?(arch)
        ["#{arch}-apple-darwin", "tar.gz"]
      when /mswin|mingw|cygwin|windows/
        win =
          case arch
          when "x86_64" then "x86_64-pc-windows-msvc"
          when "aarch64" then "aarch64-pc-windows-msvc"
          when "i686" then "i686-pc-windows-msvc"
          else raise LauncherError, "unsupported Windows arch: #{arch}"
          end
        [win, "zip"]
      when /linux/
        libc = musl? ? "musl" : "gnu"
        suffix = arch == "arm" ? "eabihf" : ""
        ["#{arch}-unknown-linux-#{libc}#{suffix}", "tar.gz"]
      else
        raise LauncherError, "unsupported OS: #{host_os}"
      end
    end

    def musl?
      return true if RbConfig::CONFIG["host_os"].downcase.include?("musl")
      Dir.glob("/lib/ld-musl-*.so.1").any?
    rescue StandardError
      false
    end

    def cache_dir
      base =
        if Gem.win_platform?
          ENV["LOCALAPPDATA"] || File.join(Dir.home, "AppData", "Local")
        else
          ENV["XDG_CACHE_HOME"] || File.join(Dir.home, ".cache")
        end
      File.join(base, "socket-patch", "bin")
    end

    # ── download + verify + extract ───────────────────────────────────────────

    def download_binary(ver, target, ext, dest)
      archive = "#{BINARY}-#{target}.#{ext}"
      base = "https://github.com/#{REPO}/releases/download/v#{ver}"

      Dir.mktmpdir("socket-patch") do |tmp|
        archive_path = File.join(tmp, archive)
        fetch("#{base}/#{archive}", archive_path)

        sums = fetch_string("#{base}/SHA256SUMS")
        verify_sha256!(archive_path, archive, sums)

        extract(archive_path, ext, tmp)
        exe = BINARY + (ext == "zip" ? ".exe" : "")
        extracted = File.join(tmp, exe)
        unless File.file?(extracted)
          raise LauncherError, "release archive #{archive} did not contain #{exe}"
        end

        FileUtils.mkdir_p(File.dirname(dest))
        FileUtils.cp(extracted, dest)
        File.chmod(0o755, dest) unless Gem.win_platform?
      end
    end

    # Follow redirects (GitHub release downloads redirect to a CDN) and stream
    # the body to `dest`.
    def fetch(url, dest, redirects = 10)
      raise LauncherError, "too many redirects fetching #{url}" if redirects.zero?
      uri = URI(url)
      Net::HTTP.start(uri.host, uri.port, use_ssl: uri.scheme == "https") do |http|
        http.request(Net::HTTP::Get.new(uri)) do |res|
          case res
          when Net::HTTPRedirection
            return fetch(res["location"], dest, redirects - 1)
          when Net::HTTPSuccess
            File.open(dest, "wb") { |f| res.read_body { |chunk| f.write(chunk) } }
          else
            raise LauncherError, "download failed (#{res.code}) for #{url}"
          end
        end
      end
    end

    def fetch_string(url, redirects = 10)
      raise LauncherError, "too many redirects fetching #{url}" if redirects.zero?
      uri = URI(url)
      res = Net::HTTP.start(uri.host, uri.port, use_ssl: uri.scheme == "https") do |http|
        http.request(Net::HTTP::Get.new(uri))
      end
      case res
      when Net::HTTPRedirection then fetch_string(res["location"], redirects - 1)
      when Net::HTTPSuccess then res.body
      else raise LauncherError, "download failed (#{res.code}) for #{url}"
      end
    end

    # SHA256SUMS lines are "<hex>  <filename>" (some tools prefix the name
    # with `*` for binary mode); match either.
    def verify_sha256!(path, archive, sums)
      expected = nil
      sums.each_line do |line|
        hex, name = line.split(/\s+/, 2)
        next unless name
        name = name.strip.sub(/\A\*/, "")
        if name == archive
          expected = hex.strip
          break
        end
      end
      raise LauncherError, "no SHA256SUMS entry for #{archive}" unless expected
      actual = Digest::SHA256.file(path).hexdigest
      return if actual.casecmp?(expected)
      raise LauncherError, "checksum mismatch for #{archive} (expected #{expected}, got #{actual})"
    end

    def extract(archive_path, ext, dir)
      ok =
        if ext == "zip"
          # bsdtar (the `tar` on modern Windows) extracts zip; fall back to
          # PowerShell Expand-Archive.
          system("tar", "-xf", archive_path, "-C", dir) ||
            system("powershell", "-NoProfile", "-Command",
                   "Expand-Archive -Force -LiteralPath '#{archive_path}' -DestinationPath '#{dir}'")
        else
          system("tar", "xzf", archive_path, "-C", dir)
        end
      raise LauncherError, "failed to extract #{File.basename(archive_path)}" unless ok
    end
  end
end

File: maven/socket-patch/src/main/java/dev/socket/socketpatch/Launcher.java
package dev.socket.socketpatch;

import java.io.IOException;
import java.io.InputStream;
import java.net.InetSocketAddress;
import java.net.ProxySelector;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.nio.file.DirectoryStream;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.nio.file.StandardCopyOption;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Comparator;
import java.util.List;
import java.util.Locale;
import java.util.regex.Pattern;
import java.util.stream.Stream;
import java.util.zip.ZipEntry;
import java.util.zip.ZipInputStream;

/**
 * Resolves and runs the prebuilt {@code socket-patch} binary for the host
 * platform (Maven Central distribution of the socket-patch CLI).
 *
 * <p>Strategy (mirrors scripts/install.sh's target mapping and the RubyGems /
 * Composer launchers):
 * <ol>
 *   <li>honor {@code SOCKET_PATCH_BIN} if it points at an executable (airgap
 *       escape);</li>
 *   <li>else use a cached binary under the per-user cache, keyed by
 *       version + target;</li>
 *   <li>else download {@code socket-patch-<target>.{tar.gz,zip}} from the
 *       matching GitHub release, verify its SHA-256 against the release's
 *       SHA256SUMS, extract the binary, cache it, and run it.</li>
 * </ol>
 */
public final class Launcher {
    /**
     * Fallback version, used ONLY when the jar manifest's
     * Implementation-Version is unavailable (e.g. running unpacked classes
     * straight from a checkout). Kept current by scripts/version-sync.sh.
     * In a real Maven-resolved jar the download uses the jar's own stamped
     * version — see {@link #version()}.
     */
    private static final String VERSION = "3.3.0";

    private static final String REPO = "SocketDev/socket-patch";
    private static final String BINARY = "socket-patch";

    /**
     * Plain release versions look like 3.3.0 — anchored full match, because a
     * suffixed version (3.3.1-SNAPSHOT, prereleases) has no matching GitHub
     * release binary and must fall back to {@link #VERSION} instead of
     * guaranteeing a 404.
     */
    private static final Pattern RELEASE_VERSION = Pattern.compile("^\\d+\\.\\d+\\.\\d+$");

    /**
     * Follow redirects (GitHub release downloads redirect to a CDN), but only
     * to HTTPS targets: {@code Redirect.NORMAL} is documented to always
     * redirect "except from HTTPS URLs to HTTP URLs", so every hop is
     * JDK-vetted to stay on HTTPS. That matters because a redirect to
     * http:// would let a network attacker serve a malicious binary AND a
     * matching SHA256SUMS (both attacker-controlled), defeating the checksum
     * check. The initial URL is separately asserted HTTPS in
     * {@link #httpsRequest(String)}.
     */
    private static final HttpClient HTTP = buildHttpClient();

    private static HttpClient buildHttpClient() {
        HttpClient.Builder builder = HttpClient.newBuilder()
                .followRedirects(HttpClient.Redirect.NORMAL);
        ProxySelector proxy = envProxySelector();
        if (proxy != null) {
            builder.proxy(proxy);
        }
        return builder.build();
    }

    /**
     * The JVM's default proxy selector only honors {@code -Dhttps.proxyHost}
     * -style system properties, never the {@code https_proxy}/{@code
     * HTTPS_PROXY} environment variables that every sibling launcher (curl in
     * install.sh, Ruby's Net::HTTP, PHP's libcurl, .NET's HttpClient) picks up
     * — so behind an env-configured egress proxy the first-run download would
     * fail only for the Maven distribution. Honor the env vars here; explicit
     * JVM proxy properties still take precedence (returning null keeps the
     * default selector, which reads them).
     */
    private static ProxySelector envProxySelector() {
        if (System.getProperty("https.proxyHost") != null
                || System.getProperty("http.proxyHost") != null) {
            return null;
        }
        for (String name : new String[] {"https_proxy", "HTTPS_PROXY", "all_proxy", "ALL_PROXY"}) {
            String value = System.getenv(name);
            if (value == null || value.isEmpty()) {
                continue;
            }
            URI uri = URI.create(value.contains("://") ? value : "http://" + value);
            if (uri.getHost() == null) {
                continue;
            }
            int port = uri.getPort() != -1
                    ? uri.getPort()
                    : ("https".equalsIgnoreCase(uri.getScheme()) ? 443 : 80);
            return ProxySelector.of(new InetSocketAddress(uri.getHost(), port));
        }
        return null;
    }

    private Launcher() {
    }

    /**
     * Entry point: resolves the platform binary, runs it with the given
     * arguments (inheriting stdio), and exits with the child's exit code.
     *
     * @param args CLI arguments passed through to the socket-patch binary
     */
    public static void main(String[] args) {
        try {
            String bin = resolveBinary();
            // Java has no exec() that replaces the process; spawn with
            // inherited stdio and propagate the child's exit status.
            List<String> cmd = new ArrayList<>();
            cmd.add(bin);
            cmd.addAll(Arrays.asList(args));
            Process child = new ProcessBuilder(cmd).inheritIO().start();
            System.exit(child.waitFor());
        } catch (LauncherError e) {
            System.err.println("socket-patch: " + e.getMessage());
            System.exit(1);
        } catch (IOException e) {
            System.err.println("socket-patch: " + e.getMessage());
            System.exit(1);
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
            System.err.println("socket-patch: interrupted");
            System.exit(1);
        }
    }

    /** Launcher-level failure with a user-facing message. */
    private static final class LauncherError extends RuntimeException {
        private static final long serialVersionUID = 1L;

        LauncherError(String message) {
            super(message);
        }
    }

    // ── binary resolution ────────────────────────────────────────────────────

    private static String resolveBinary() {
        String env = System.getenv("SOCKET_PATCH_BIN");
        if (env != null && !env.isEmpty()) {
            Path p = Paths.get(env);
            if (Files.isRegularFile(p) && Files.isExecutable(p)) {
                return env;
            }
        }

        String ver = version();
        String[] targetExt = detectTarget();
        String target = targetExt[0];
        String ext = targetExt[1];
        String exe = BINARY + (isWindows() ? ".exe" : "");
        Path cached = cacheDir().resolve(ver).resolve(target).resolve(exe);
        // Cache hit: the cached binary was SHA-256-verified when first
        // downloaded and lives under the user's own cache dir. We trust it
        // without re-verifying (re-verification would require re-fetching
        // SHA256SUMS every run), matching npx / pip / rustup; an attacker able
        // to write here can already replace the installed jar or the binary
        // itself.
        if (Files.isRegularFile(cached) && Files.isExecutable(cached)) {
            return cached.toString();
        }

        downloadBinary(ver, target, ext, cached);
        return cached.toString();
    }

    /**
     * The version to fetch — the binary MUST match the artifact the user
     * actually resolved, so derive it from the jar's own Implementation-Version
     * manifest attribute (stamped from {@code project.version} by
     * maven-jar-plugin) rather than trusting the {@code VERSION} constant
     * (which version-sync.sh keeps current but which could drift). Falls back
     * to the constant when the manifest isn't available (e.g. running unpacked
     * classes from a checkout) or reports a non-release version with no
     * matching release binary.
     */
    private static String version() {
        Package pkg = Launcher.class.getPackage();
        String v = pkg == null ? null : pkg.getImplementationVersion();
        if (v != null && RELEASE_VERSION.matcher(v).matches()) {
            return v;
        }
        return VERSION;
    }

    /**
     * Map the host to a release target triple + archive extension. Mirrors
     * scripts/install.sh.
     */
    private static String[] detectTarget() {
        String osName = System.getProperty("os.name", "").toLowerCase(Locale.ROOT);
        String osArch = System.getProperty("os.arch", "").toLowerCase(Locale.ROOT);

        String arch;
        if (osArch.matches(".*(x86_64|x64|amd64).*")) {
            arch = "x86_64";
        } else if (osArch.matches(".*(aarch64|arm64).*")) {
            arch = "aarch64";
        } else if (osArch.matches(".*(i[3-6]86|x86).*")) {
            arch = "i686";
        } else if (osArch.matches(".*(armv7|armhf|arm\\b).*")) {
            arch = "arm";
        } else {
            throw new LauncherError("unsupported CPU architecture: " + osArch);
        }

        // Check macOS before Windows: "darwin" contains "win".
        if (osName.contains("mac") || osName.contains("darwin")) {
            if (!arch.equals("x86_64") && !arch.equals("aarch64")) {
                throw new LauncherError("unsupported macOS arch: " + arch);
            }
            return new String[] {arch + "-apple-darwin", "tar.gz"};
        }
        if (osName.startsWith("windows")) {
            if (arch.equals("x86_64") || arch.equals("aarch64") || arch.equals("i686")) {
                return new String[] {arch + "-pc-windows-msvc", "zip"};
            }
            throw new LauncherError("unsupported Windows arch: " + arch);
        }
        if (osName.contains("linux")) {
            String libc = isMusl() ? "musl" : "gnu";
            String suffix = arch.equals("arm") ? "eabihf" : "";
            return new String[] {arch + "-unknown-linux-" + libc + suffix, "tar.gz"};
        }
        throw new LauncherError("unsupported OS: " + osName);
    }

    private static boolean isMusl() {
        try (DirectoryStream<Path> ds =
                Files.newDirectoryStream(Paths.get("/lib"), "ld-musl-*.so.1")) {
            return ds.iterator().hasNext();
        } catch (IOException e) {
            return false;
        }
    }

    private static boolean isWindows() {
        return System.getProperty("os.name", "").toLowerCase(Locale.ROOT).startsWith("windows");
    }

    private static Path cacheDir() {
        String base;
        if (isWindows()) {
            String localAppData = System.getenv("LOCALAPPDATA");
            base = (localAppData != null && !localAppData.isEmpty())
                    ? localAppData
                    : Paths.get(System.getProperty("user.home"), "AppData", "Local").toString();
        } else {
            String xdg = System.getenv("XDG_CACHE_HOME");
            base = (xdg != null && !xdg.isEmpty())
                    ? xdg
                    : Paths.get(System.getProperty("user.home"), ".cache").toString();
        }
        return Paths.get(base, "socket-patch", "bin");
    }

    // ── download + verify + extract ──────────────────────────────────────────

    private static void downloadBinary(String ver, String target, String ext, Path dest) {
        String archive = BINARY + "-" + target + "." + ext;
        String base = "https://github.com/" + REPO + "/releases/download/v" + ver;

        Path tmp;
        try {
            tmp = Files.createTempDirectory("socket-patch");
        } catch (IOException e) {
            throw new LauncherError("could not create temp dir: " + e.getMessage());
        }
        try {
            Path archivePath = tmp.resolve(archive);
            fetch(base + "/" + archive, archivePath);

            String sums = fetchString(base + "/SHA256SUMS");
            verifySha256(archivePath, archive, sums);

            extract(archivePath, ext, tmp);
            String exe = BINARY + (ext.equals("zip") ? ".exe" : "");
            Path extracted = tmp.resolve(exe);
            if (!Files.isRegularFile(extracted)) {
                throw new LauncherError("release archive " + archive + " did not contain " + exe);
            }

            try {
                Files.createDirectories(dest.getParent());
                Files.copy(extracted, dest, StandardCopyOption.REPLACE_EXISTING);
            } catch (IOException e) {
                throw new LauncherError("could not cache binary at " + dest + ": " + e.getMessage());
            }
            if (!isWindows()) {
                dest.toFile().setExecutable(true, false);
            }
        } finally {
            deleteRecursively(tmp); // best-effort temp cleanup (Ruby's mktmpdir block equivalent)
        }
    }

    /**
     * Require HTTPS for every request — including after a redirect. The
     * shared client's {@code Redirect.NORMAL} policy never follows an
     * HTTPS-to-HTTP redirect (see {@link #HTTP}); this asserts the INITIAL
     * URL is HTTPS too, so no request ever leaves over plain HTTP. A
     * non-HTTPS URL anywhere would let a network attacker serve a malicious
     * binary AND a matching SHA256SUMS (both attacker-controlled), defeating
     * the checksum check.
     */
    private static HttpRequest httpsRequest(String url) {
        if (!url.startsWith("https://")) {
            throw new LauncherError("refusing non-HTTPS URL: " + url);
        }
        return HttpRequest.newBuilder(URI.create(url)).GET().build();
    }

    private static void fetch(String url, Path dest) {
        HttpResponse<Path> res;
        try {
            res = HTTP.send(httpsRequest(url), HttpResponse.BodyHandlers.ofFile(dest));
        } catch (IOException e) {
            throw new LauncherError("download failed for " + url + ": " + e.getMessage());
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
            throw new LauncherError("download interrupted for " + url);
        }
        if (res.statusCode() / 100 != 2) {
            throw new LauncherError("download failed (" + res.statusCode() + ") for " + url);
        }
    }

    private static String fetchString(String url) {
        HttpResponse<String> res;
        try {
            res = HTTP.send(httpsRequest(url), HttpResponse.BodyHandlers.ofString());
        } catch (IOException e) {
            throw new LauncherError("download failed for " + url + ": " + e.getMessage());
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
            throw new LauncherError("download interrupted for " + url);
        }
        if (res.statusCode() / 100 != 2) {
            throw new LauncherError("download failed (" + res.statusCode() + ") for " + url);
        }
        return res.body();
    }

    /**
     * SHA256SUMS lines are {@code "<hex>  <filename>"} (some tools prefix the
     * name with {@code *} for binary mode); match either.
     */
    private static void verifySha256(Path path, String archive, String sums) {
        String expected = null;
        for (String line : sums.split("\\r?\\n")) {
            String[] parts = line.trim().split("\\s+", 2);
            if (parts.length < 2) {
                continue;
            }
            String name = parts[1].trim();
            if (name.startsWith("*")) {
                name = name.substring(1);
            }
            if (name.equals(archive)) {
                expected = parts[0];
                break;
            }
        }
        if (expected == null) {
            throw new LauncherError("no SHA256SUMS entry for " + archive);
        }
        String actual = sha256Hex(path);
        if (!actual.equalsIgnoreCase(expected)) {
            throw new LauncherError(
                    "checksum mismatch for " + archive + " (expected " + expected + ", got " + actual + ")");
        }
    }

    private static String sha256Hex(Path file) {
        MessageDigest md;
        try {
            md = MessageDigest.getInstance("SHA-256");
        } catch (NoSuchAlgorithmException e) {
            throw new LauncherError("SHA-256 unavailable: " + e.getMessage());
        }
        try (InputStream in = Files.newInputStream(file)) {
            byte[] buf = new byte[65536];
            int n;
            while ((n = in.read(buf)) != -1) {
                md.update(buf, 0, n);
            }
        } catch (IOException e) {
            throw new LauncherError("could not read " + file + ": " + e.getMessage());
        }
        StringBuilder sb = new StringBuilder();
        for (byte b : md.digest()) {
            sb.append(String.format("%02x", b));
        }
        return sb.toString();
    }

    private static void extract(Path archivePath, String ext, Path dir) {
        if (ext.equals("zip")) {
            extractZip(archivePath, dir);
            return;
        }
        // Shell out to tar for tar.gz — the same choice as the Ruby and PHP
        // launchers: the JDK has no built-in tar support, tar ships on every
        // supported non-Windows platform, and the archive's checksum was
        // already verified against SHA256SUMS before extraction.
        try {
            Process p = new ProcessBuilder("tar", "xzf", archivePath.toString(), "-C", dir.toString())
                    .inheritIO()
                    .start();
            if (p.waitFor() != 0) {
                throw new LauncherError("failed to extract " + archivePath.getFileName());
            }
        } catch (IOException e) {
            throw new LauncherError(
                    "failed to extract " + archivePath.getFileName() + ": " + e.getMessage());
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
            throw new LauncherError("failed to extract " + archivePath.getFileName() + ": interrupted");
        }
    }

    /** Zip extraction (Windows archives) with a zip-slip guard on every entry. */
    private static void extractZip(Path archivePath, Path dir) {
        Path root = dir.toAbsolutePath().normalize();
        try (ZipInputStream zin = new ZipInputStream(Files.newInputStream(archivePath))) {
            ZipEntry entry;
            while ((entry = zin.getNextEntry()) != null) {
                // Zip-slip guard: resolve + normalize each entry path and
                // refuse anything that escapes the extraction directory.
                Path out = root.resolve(entry.getName()).normalize();
                if (!out.startsWith(root)) {
                    throw new LauncherError(
                            "refusing zip entry escaping extraction dir: " + entry.getName());
                }
                if (entry.isDirectory()) {
                    Files.createDirectories(out);
                } else {
                    if (out.getParent() != null) {
                        Files.createDirectories(out.getParent());
                    }
                    Files.copy(zin, out, StandardCopyOption.REPLACE_EXISTING);
                }
                zin.closeEntry();
            }
        } catch (IOException e) {
            throw new LauncherError(
                    "failed to extract " + archivePath.getFileName() + ": " + e.getMessage());
        }
    }

    private static void deleteRecursively(Path root) {
        try (Stream<Path> walk = Files.walk(root)) {
            walk.sorted(Comparator.reverseOrder()).forEach(p -> {
                try {
                    Files.deleteIfExists(p);
                } catch (IOException ignored) {
                    // best effort
                }
            });
        } catch (IOException ignored) {
            // best effort
        }
    }
}
